fuels-core/src/signature.rs

```diff
@@ -0,0 +1,195 @@
+use crate::Bits256;
+use fuel_tx::{crypto::Hasher, Bytes64};
+use fuel_types::Address;
+use fuel_vm::crypto::secp256k1_sign_compact_recover;
+use std::{convert::TryFrom, fmt, str::FromStr};
+use thiserror::Error;
+
+/// An error involving a signature.
+#[derive(Debug, Error)]
+pub enum SignatureError {
+    /// Invalid length
+    #[error("invalid signature length, got {0}, expected 64")]
+    InvalidLength(usize),
+    /// When parsing a signature from string to hex
+    #[error(transparent)]
+    DecodingError(#[from] hex::FromHexError),
+    /// Thrown when signature verification failed (i.e. when the address that
+    /// produced the signature did not match the expected address)
+    #[error("Signature verification failed. Expected {0}, got {1}")]
+    VerificationError(Address, Address),
+    /// Error in recovering public key from signature
+    #[error("Public key recovery error")]
+    RecoveryError,
+}
+
+/// Recovery message data.
+#[derive(Clone, Debug, PartialEq)]
+pub enum RecoveryMessage {
+    /// Message bytes
+    Data(Vec<u8>),
+    /// Message hash
+    Hash(Bits256),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+/// An ECDSA signature. Encoded as:
+///
+/// ```plaintext
+/// |    32 bytes   ||           32 bytes           |
+/// [256-bit r value][1-bit v value][255-bit s value]
+/// ```
+///
+/// The encoding of the signature was derived from
+/// [Compact Signature Representation](https://eips.ethereum.org/EIPS/eip-2098).
+///
+/// Signatures are represented as the `r` and `s` (each 32 bytes),
+/// and `v` (1-bit) values of the signature. `r` and `s` take on
+/// their usual meaning while `v` is used for recovering the public
+/// key from a signature more quickly.
+pub struct Signature {
+    pub compact: Bytes64,
+}
+
+impl fmt::Display for Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(self.compact))
+    }
+}
+
+impl Signature {
+    /// Verifies that signature on `message` was produced by `address`
+    pub fn verify<M, A>(&self, message: M, address: A) -> Result<(), SignatureError>
+    where
+        M: Into<RecoveryMessage>,
+        A: Into<Address>,
+    {
+        let address = address.into();
+        let recovered = self.recover(message)?;
+
+        if recovered != address {
+            return Err(SignatureError::VerificationError(address, recovered));
+        }
+
+        Ok(())
+    }
+
+    /// Recovers the Ethereum address which was used to sign the given message.
+    pub fn recover<M>(&self, message: M) -> Result<Address, SignatureError>
```

Comment:
Sometimes users only have the message hash, not the original message. There should be a method to recover from a message directly (you can enforce the message is 32 bytes in that case, I'd say).

Reply:
I know. This method works both for original message and hashed message, check below. There's a match checking whether it's "Data" or "Hash".

Reply:
Ohhh now I get it. Can you document this feature in this section and the above?